src/db.c

```diff
@@ -462,90 +455,46 @@ long long emptyDb(int dbnum, int flags, void(callback)(dict*)) {
-
-    return backup;
+    return tempDb;
 }
 
-/* Discard a previously created backup, this can be slow (similar to FLUSHALL)
- * Arguments are similar to the ones of emptyDb, see EMPTYDB_ flags. */
-void discardDbBackup(dbBackup *backup, int flags, void(callback)(dict*)) {
-    int async = (flags & EMPTYDB_ASYNC);
-
-    /* Release main DBs backup . */
-    emptyDbStructure(backup->dbarray, -1, async, callback);
-    for (int i=0; i<server.dbnum; i++) {
-        dictRelease(backup->dbarray[i].dict);
-        dictRelease(backup->dbarray[i].expires);
-    }
-
-    /* Release slots to keys map backup if enable cluster. */
-    if (server.cluster_enabled) freeSlotsToKeysMap(backup->slots_to_keys, async);
-
-    /* Release backup. */
-    zfree(backup->dbarray);
-    zfree(backup);
-
-    moduleFireServerEvent(REDISMODULE_EVENT_REPL_BACKUP,
-                          REDISMODULE_SUBEVENT_REPL_BACKUP_DISCARD,
-                          NULL);
-}
+/* Discard tempDb, this can be slow (similar to FLUSHALL), but it's always async */
+ void discardTempDb(tempDb *tempDb, void(callback)(dict*)) {
+    int async = 1;
 
-/* Restore the previously created backup (discarding what currently resides
- * in the db).
- * This function should be called after the current contents of the database
- * was emptied with a previous call to emptyDb (possibly using the async mode). */
-void restoreDbBackup(dbBackup *backup) {
-    /* Restore main DBs. */
+    /* Release temp DBs */
+    emptyDbStructure(tempDb->dbarray, -1, async, callback);
     for (int i=0; i<server.dbnum; i++) {
-        serverAssert(dictSize(server.db[i].dict) == 0);
-        serverAssert(dictSize(server.db[i].expires) == 0);
-        dictRelease(server.db[i].dict);
-        dictRelease(server.db[i].expires);
-        server.db[i] = backup->dbarray[i];
+        dictRelease(tempDb->dbarray[i].dict);
+        dictRelease(tempDb->dbarray[i].expires);
     }
 
-    /* Restore slots to keys map backup if enable cluster. */
-    if (server.cluster_enabled) {
-        serverAssert(server.cluster->slots_to_keys->numele == 0);
-        raxFree(server.cluster->slots_to_keys);
-        server.cluster->slots_to_keys = backup->slots_to_keys;
-        memcpy(server.cluster->slots_keys_count, backup->slots_keys_count,
-                sizeof(server.cluster->slots_keys_count));
-    }
+    /* Release slots to keys map if enable cluster. */
+    if (server.cluster_enabled) freeSlotsToKeysMap(tempDb->slots_to_keys, async);
 
-    /* Release backup. */
-    zfree(backup->dbarray);
-    zfree(backup);
+    zfree(tempDb->dbarray);
+    zfree(tempDb);
 
-    moduleFireServerEvent(REDISMODULE_EVENT_REPL_BACKUP,
-                          REDISMODULE_SUBEVENT_REPL_BACKUP_RESTORE,
-                          NULL);
+    // TODO: adjust this
+    // moduleFireServerEvent(REDISMODULE_EVENT_REPL_BACKUP,
+    //                       REDISMODULE_SUBEVENT_REPL_BACKUP_DISCARD,
+    //                       NULL);
```

Comment:
This change is gonna break modules that use the `RedisModuleEvent_ReplBackup` mechanism anyway, so i suggest the following:

1. We deprecate `RedisModuleEvent_ReplBackup`. i.e. starting redis 7.0, we never fire that event. we do that since in the new "swapdb" mechanism there's no concept of backup and restore, instead there's a concept of temporary db.
2. We create a new API named `RedisModuleEvent_ReplAsync`, holding 3 sub-events; STARTED, COMPLETEED, ABORTED.
3. We add another module flag for `RedisModule_SetModuleOptions` with which the module can declare it supports this mechanism, i.e. REDISMODULE_OPTIONS_HANDLE_REPL_ASYNC.
4. in replication.c, If there are modules loaded which registered a data type, and didn't declare they're supporting this, we fall back to the alternative (either the `on-empty-db` diskless replication, or disk-based replication).